### PR TITLE
Cleaning up the thread safety for a couple of items in Channel.

### PR DESCRIPTION
**Pull request by jeffret-b**
If we're going to change the threading model in Channel, we should end up with something that makes sense and is coherent. Just modifying it and adding new threading models doesn't help for future clarity and maintenance.

I propose we follow through and complete the threading model upgrades / fixes you propose. It doesn't take much more and leaves us in a better state.

### CODE REVIEW

File: src/main/java/hudson/remoting/Channel.java [hunk 1]
```diff
@@ -159,7 +159,7 @@
      * Requests that are sent to the remote side for execution, yet we are waiting locally until
      * we hear back their responses.
      */
-    /*package*/ final Map<Integer,Request<? extends Serializable,? extends Throwable>> pendingCalls = new Hashtable<>();
+    /*package*/ final Map<Integer,Request<? extends Serializable,? extends Throwable>> pendingCalls = new ConcurrentHashMap<>();
```
Comment (jeffret-b) [inline src/main/java/hudson/remoting/Channel.java:162]: ConcurrentHashMap is a good model for what this needs to do. Based upon the existing implementation, it's possible it doesn't need thread safety but that seems unlikely. ConcurrentHashMap has low negative impact, but still provides thread safety if it really is needed.

File: src/main/java/hudson/remoting/Channel.java [hunk 2]
```diff
@@ -170,8 +170,7 @@
     /**
      * Records the {@link Request}s being executed on this channel, sent by the remote peer.
      */
-    /*package*/ final Map<Integer,Request<?,?>> executingCalls =
-        Collections.synchronizedMap(new Hashtable<>());
+    /*package*/ final Map<Integer,Request<?,?>> executingCalls = new ConcurrentHashMap<>();
```
Comment (jeffret-b) [inline src/main/java/hudson/remoting/Channel.java:173]: It's odd that this one has had a threading model whereas the previous hasn't. I don't know why, but it's best to be consistent and ConcurrentHashMap makes better sense.

File: src/main/java/hudson/remoting/Channel.java [hunk 3]
```diff
@@ -1080,18 +1079,17 @@ public void terminate(@Nonnull IOException e) {
                     logger.log(Level.WARNING, "Failed to close down the reader side of the transport", x);
                 }
                 try {
-                    synchronized (pendingCalls) {
-                        for (Request<?, ?> req : pendingCalls.values())
-                            req.abort(e);
-                        pendingCalls.clear();
+                    for (Request<?, ?> req : pendingCalls.values()) {
```
Comment (jeffret-b) [inline src/main/java/hudson/remoting/Channel.java:1082]: The iterator provided here by the ConcurrentHashMap will be consistent and coherent over the execution of this for loop. Synchronization on the object, where it isn't done elsewhere, is a broken, mixed threading model.

It's possible that something could be added to the map between the time the iterator is created and the following clear() operation is performed. We could additional logic to correctly deal with this. However, this logic is inside a terminate() method. If we miss aborting a few requests, that shouldn't make much difference as the channel is terminating anyway.

If we do need to deal with that threading inconsistency, we need something different than what is currently released.

File: src/main/java/hudson/remoting/Channel.java [hunk 4]
```diff
@@ -1080,18 +1079,17 @@ public void terminate(@Nonnull IOException e) {
...
+                        req.abort(e);
                     }
-                    synchronized (executingCalls) {
-                        for (Request<?, ?> r : executingCalls.values()) {
-                            java.util.concurrent.Future<?> f = r.future;
-                            if (f != null) f.cancel(true);
+                    pendingCalls.clear();
```
Comment (jeffret-b) [inline src/main/java/hudson/remoting/Channel.java:1085]: Similar comments here as for pendingCalls. Mixing broken, inconsistent threading models doesn't help.

File: src/main/java/hudson/remoting/Channel.java [hunk 5]
```diff
@@ -1443,7 +1441,6 @@ public void dumpDiagnostics(@Nonnull PrintWriter w) throws IOException {
         w.printf("  Last command sent=%s%n", new Date(lastCommandSentAt));
         w.printf("  Last command received=%s%n", new Date(lastCommandReceivedAt));
 
-        // TODO: Synchronize when Hashtable gets replaced by a modern collection.
```
Comment (jeffret-b) [inline src/main/java/hudson/remoting/Channel.java:1446]: Replacing it with a modern collection, that offers sufficient threading safety for obtaining the size.

**Comment by jeffret-b** [comment]
@jvz  take a look at this to go with a single, consistent threading model if we're going to start changing something in the area.

**Review by jvz — APPROVED** [review]
All good ideas, thanks!

**Comment by jvz** [comment]
Can you resolve this merge conflict? I have no idea where it came from. You can just apply your changes to my PR directly.

**Pull request closed without merging**